plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotPageSourceProvider.java

```diff
@@ -113,4 +143,279 @@ public ConnectorPageSource createPageSource(
+        {
+            this.pinotGrpcClient = requireNonNull(pinotGrpcClient, "pinotGrpcClient is null");
+            this.split = requireNonNull(split, "split is null");
+            this.query = requireNonNull(query, "query is null");
+            this.rowCountChecker = requireNonNull(rowCountChecker, "rowCountChecker is null");
+        }
+
+        @Override
+        public long getReadTimeNanos()
+        {
+            return readTimeNanos;
+        }
+
+        @Override
+        public long getMemoryUsageBytes()
+        {
+            return 0;
+        }
+
+        @Override
+        public boolean endOfData()
+        {
+            return !responseIterator.hasNext();
+        }
+
+        @Override
+        public boolean isDataFetched()
+        {
+            return isPinotDataFetched;
+        }
+
+        @Override
+        public void fetchData()
+        {
+            long startTimeNanos = System.nanoTime();
+            this.responseIterator = queryPinot();
+            readTimeNanos += System.nanoTime() - startTimeNanos;
+            isPinotDataFetched = true;
+        }
+
+        @Override
+        public PinotDataTableWithSize getNextDataTable()
+        {
+            PinotDataTableWithSize dataTableWithSize = responseIterator.next();
+            rowCountChecker.checkTooManyRows(dataTableWithSize.getDataTable());
+            checkExceptions(dataTableWithSize.getDataTable(), split, query);
+            return dataTableWithSize;
+        }
+
+        private ResponseIterator queryPinot()
+        {
+            String serverHost = split.getSegmentHost().orElseThrow(() -> new PinotException(PinotErrorCode.PINOT_INVALID_PQL_GENERATED, Optional.empty(), "Expected the segment split to contain the host"));
+            return new ResponseIterator(pinotGrpcClient.submitQuery(query, serverHost, split.getSegments()));
+        }
+    }
+
+    public static class ResponseIterator
+            extends AbstractIterator<PinotDataTableWithSize>
+    {
+        private final Iterator<Server.ServerResponse> responseIterator;
+
+        public ResponseIterator(Iterator<Server.ServerResponse> responseIterator)
+        {
+            this.responseIterator = requireNonNull(responseIterator, "responseIterator is null");
+        }
+
+        @Override
+        protected PinotDataTableWithSize computeNext()
+        {
+            if (!responseIterator.hasNext()) {
+                return endOfData();
+            }
+            Server.ServerResponse response = responseIterator.next();
+            String responseType = response.getMetadataMap().get(CommonConstants.Query.Response.MetadataKeys.RESPONSE_TYPE);
+            if (responseType.equals(CommonConstants.Query.Response.ResponseType.METADATA)) {
+                return endOfData();
+            }
+            ByteBuffer buffer = response.getPayload().asReadOnlyByteBuffer();
```

Comment:
Should this memory be tracked in the PageSource?

Reply:
Yep! Added it.